src/main/java/com/cleanroommc/modularui/value/sync/GenericListSyncHandler.java

```diff
@@ -10,108 +11,80 @@
-    }
-
-    public GenericListSyncHandler(@NotNull Supplier<List<T>> getter,
-                                  @Nullable Consumer<List<T>> setter,
-                                  @NotNull IByteBufDeserializer<T> deserializer,
-                                  @NotNull IByteBufSerializer<T> serializer,
-                                  @Nullable IEquals<T> equals) {
-        this.getter = Objects.requireNonNull(getter);
-        this.cache.addAll(getter.get());
-        this.setter = setter;
-        this.deserializer = deserializer;
-        this.serializer = serializer;
-        this.equals = equals != null ? IEquals.wrapNullSafe(equals) : Objects::equals;
+    public GenericListSyncHandler(@NotNull Supplier<List<T>> getter, @Nullable Consumer<List<T>> setter, @NotNull IByteBufDeserializer<T> deserializer, @NotNull IByteBufSerializer<T> serializer, @Nullable IEquals<T> equals, @Nullable ICopy<T> copy) {
+        super(getter, setter, deserializer, serializer, equals, copy);
+        setCache(getter.get());
     }
 
     @Override
-    public void setValue(List<T> value, boolean setSource, boolean sync) {
+    protected void setCache(List<T> value) {
         this.cache.clear();
-        this.cache.addAll(value);
-        if (setSource && this.setter != null) {
-            this.setter.accept(value);
-        }
-        if (sync) {
-            sync(0, this::write);
+        for (T item : value) {
+            this.cache.add(copyValue(item));
         }
     }
 
     @Override
-    public boolean updateCacheFromSource(boolean isFirstSync) {
-        List<T> i = this.getter.get();
-        if (isFirstSync || didValuesChange(i)) {
-            setValue(i, false, false);
-            return true;
-        }
-        return false;
-    }
-
-    private boolean didValuesChange(List<T> newValues) {
+    protected boolean didValuesChange(List<T> newValues) {
         if (this.cache.size() != newValues.size()) return true;
-        for (int i = 0, n = newValues.size(); i < n; i++) {
-            if (!this.equals.areEqual(this.cache.get(i), newValues.get(i))) {
-                return true;
-            }
+        for (int i = 0; i < this.cache.size(); i++) {
+            if (!areValuesEqual(this.cache.get(i), newValues.get(i))) return true;
         }
         return false;
     }
 
     @Override
-    public void write(PacketBuffer buffer) throws IOException {
-        buffer.writeVarIntToBuffer(this.cache.size());
-        for (T t : this.cache) {
-            this.serializer.serialize(buffer, t);
-        }
+    public List<T> getValue() {
+        return Collections.unmodifiableList(this.cache);
     }
 
     @Override
     public void read(PacketBuffer buffer) throws IOException {
         this.cache.clear();
-        for (int i = 0, n = buffer.readVarIntFromBuffer(); i < n; i++) {
-            this.cache.add(this.deserializer.deserialize(buffer));
+        for (int i = 0; i < buffer.readVarIntFromBuffer(); i++) {
```

Comment:
Here and in other `GenericCollectionSyncHandler` implementations you are erroneously reading the size in each iteration, which overflows the buffer.